File: app-miyou/src/main/java/com/miyou/app/domain/credit/model/CreditTransaction.java
package com.miyou.app.domain.credit.model;

import java.time.Instant;

import com.miyou.app.domain.dialogue.model.UserId;

public record CreditTransaction(
	CreditTransactionId transactionId,
	UserId userId,
	CreditTransactionType type,
	CreditSource source,
	long amount,
	long balanceBefore,
	long balanceAfter,
	String referenceId,
	Instant createdAt
) {
	public CreditTransaction {
		if (transactionId == null) {
			throw new IllegalArgumentException("transactionId cannot be null");
		}
		if (userId == null) {
			throw new IllegalArgumentException("userId cannot be null");
		}
		if (type == null) {
			throw new IllegalArgumentException("type cannot be null");
		}
		if (source == null) {
			throw new IllegalArgumentException("source cannot be null");
		}
		if (amount <= 0) {
			throw new IllegalArgumentException("amount must be positive");
		}
		if (createdAt == null) {
			createdAt = Instant.now();
		}
	}

	public static CreditTransaction of(UserId userId,
		CreditTransactionType type,
		CreditSource source,
		long amount,
		long balanceBefore,
		long balanceAfter) {
		return new CreditTransaction(
			CreditTransactionId.generate(),
			userId,
			type,
			source,
			amount,
			balanceBefore,
			balanceAfter,
			null,
			Instant.now());
	}

	public static CreditTransaction of(UserId userId,
		CreditTransactionType type,
		CreditSource source,
		long amount,
		long balanceBefore,
		long balanceAfter,
		String referenceId) {
		return new CreditTransaction(
			CreditTransactionId.generate(),
			userId,
			type,
			source,
			amount,
			balanceBefore,
			balanceAfter,
			referenceId,
			Instant.now());
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/domain/credit/model/CreditTransaction.kt
package com.miyou.app.domain.credit.model

import com.miyou.app.domain.dialogue.model.UserId
import java.time.Instant

data class CreditTransaction(
    val transactionId: CreditTransactionId,
    val userId: UserId,
    val type: CreditTransactionType,
    val source: CreditSource,
    val amount: Long,
    val balanceBefore: Long,
    val balanceAfter: Long,
    val referenceId: String?,
    val createdAt: Instant? = null,
) {
    init {
        require(amount > 0) { "amount must be positive" }
    }

    fun withCreatedAtOrNow(): CreditTransaction = if (createdAt == null) copy(createdAt = Instant.now()) else this

    companion object {
        @JvmStatic
        fun of(
            userId: UserId,
            type: CreditTransactionType,
            source: CreditSource,
            amount: Long,
            balanceBefore: Long,
            balanceAfter: Long,
        ): CreditTransaction =
            CreditTransaction(
                CreditTransactionId.generate(),
                userId,
                type,
                source,
                amount,
                balanceBefore,
                balanceAfter,
                referenceId = null,
            ).withCreatedAtOrNow()

        @JvmStatic
        fun of(
            userId: UserId,
            type: CreditTransactionType,
            source: CreditSource,
            amount: Long,
            balanceBefore: Long,
            balanceAfter: Long,
            referenceId: String,
        ): CreditTransaction =
            CreditTransaction(
                CreditTransactionId.generate(),
                userId,
                type,
                source,
                amount,
                balanceBefore,
                balanceAfter,
                referenceId = referenceId,
            ).withCreatedAtOrNow()
    }

    fun transactionId(): CreditTransactionId = transactionId

    fun userId(): UserId = userId

    fun type(): CreditTransactionType = type

    fun source(): CreditSource = source

    fun amount(): Long = amount

    fun balanceBefore(): Long = balanceBefore

    fun balanceAfter(): Long = balanceAfter

    fun referenceId(): String? = referenceId

    fun createdAt(): java.time.Instant? = createdAt
}
